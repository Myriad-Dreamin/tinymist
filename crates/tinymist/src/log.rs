//! Logging Functionality

use serde::{Deserialize, Serialize};

use super::*;

/// Options for initializing the logger.
pub struct InitLogOpts {
    /// Whether the command is transient (e.g., compile).   
    pub is_transient_cmd: bool,
    /// Whether the command is a test without verbose output.
    pub is_test_no_verbose: bool,
    /// Redirects output via LSP/DAP notification.
    pub output: Option<LspClient>,
}

/// Initializes the logger for the Tinymist library.
pub fn init_log(
    InitLogOpts {
        is_transient_cmd,
        is_test_no_verbose,
        output,
    }: InitLogOpts,
) -> anyhow::Result<()> {
    use log::LevelFilter::*;

    let base_no_info = is_transient_cmd || is_test_no_verbose;
    let base_level = if base_no_info { Warn } else { Info };
    let preview_level = if is_test_no_verbose { Warn } else { Debug };
    let diag_level = if is_test_no_verbose { Warn } else { Info };

    let mut builder = env_logger::builder();
    if let Some(output) = output {
        builder.target(LogNotification::create(output));
    }

    // In WebAssembly, we use a custom notification for logging.
    #[cfg(target_arch = "wasm32")]
    {
        builder.format(|f, record| {
            use std::io::Write;
            let ts = tinymist_std::time::utc_now();

            write!(f, "[")?;
            ts.format_into(f, &tinymist_std::time::Rfc3339)
                .map_err(std::io::Error::other)?;
            writeln!(
                f,
                " {level:<5} {module_path} {file_path}:{line} {target}] {args}",
                level = record.level(),
                module_path = record.module_path().unwrap_or("unknown"),
                file_path = record.file().unwrap_or("unknown"),
                line = record.line().unwrap_or(0),
                target = record.target(),
                args = record.args()
            )
        });
    }

    Ok(builder
        .filter_module("tinymist", base_level)
        .filter_module("tinymist_preview", preview_level)
        .filter_module("typlite", base_level)
        .filter_module("reflexo", base_level)
        .filter_module("sync_ls", base_level)
        .filter_module("reflexo_typst2vec::pass::span2vec", Error)
        .filter_module("reflexo_typst::diag::console", diag_level)
        .try_init()?)
}

struct LogNotification(LspClient, Vec<u8>);

impl LogNotification {
    /// Creates a new `LogNotification` with the given LSP client and an empty
    /// buffer.
    fn create(output: LspClient) -> env_logger::Target {
        env_logger::Target::Pipe(Box::new(Self(output, vec![])))
    }
}

impl std::io::Write for LogNotification {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        if buf.is_empty() {
            return Ok(0);
        }

        self.1.extend_from_slice(buf);
        Ok(buf.len())
    }

    // todo: the from_utf8_lossy may break non-ascii characters and inefficient
    fn flush(&mut self) -> std::io::Result<()> {
        let data = String::from_utf8_lossy(self.1.as_slice()).to_string();
        self.1.clear();
        self.0.send_notification::<Log>(&Log { data });
        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Log {
    data: String,
}

impl lsp_types::notification::Notification for Log {
    const METHOD: &'static str = "tmLog";
    type Params = Self;
}
