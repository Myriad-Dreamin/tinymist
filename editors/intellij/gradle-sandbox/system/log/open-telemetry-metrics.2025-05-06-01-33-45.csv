# OpenTelemetry Metrics report: .csv, 4 fields: 
# <metric name>, <period start, nanoseconds>, <period end, nanoseconds>, <metric value>
# See CsvMetricsExporter for details.
# 
# NAME, PERIOD_START_NANOS, PERIOD_END_NANOS, VALUE
Indexes.entries.lookups,1746488025455000000,1746488037605000000,1.0
Indexes.allKeys.lookups,1746488025455000000,1746488037605000000,0.0
Indexes.stubs.lookupDuration90PMs,1746488025455000000,1746488037605000000,0.0
Indexes.entries.lookupDuration90PMs,1746488025455000000,1746488037605000000,3.0
Indexes.stubs.lookupDurationMaxMs,1746488025455000000,1746488037605000000,0.0
Indexes.entries.lookupDurationMaxMs,1746488025455000000,1746488037605000000,3.0
Indexes.stubs.lookups,1746488025455000000,1746488037605000000,0.0
Indexes.allKeys.lookupDuration90PMs,1746488025455000000,1746488037605000000,0.0
Indexes.stubs.lookupDurationAvgMs,1746488025455000000,1746488037605000000,0.0
Indexes.allKeys.lookupDurationAvgMs,1746488025455000000,1746488037605000000,0.0
Indexes.entries.lookupDurationAvgMs,1746488025455000000,1746488037605000000,3.0
Indexes.allKeys.lookupDurationMaxMs,1746488025455000000,1746488037605000000,0.0
workspaceModel.updates.precise.ms,1746488025455000000,1746488037605000000,0
workspaceModel.check.recursive.update.ms,1746488025455000000,1746488037605000000,0
workspaceModel.loading.total.ms,1746488025455000000,1746488037605000000,25
workspaceModel.update.unloaded.entities.ms,1746488025455000000,1746488037605000000,0
workspaceModel.collect.changes.ms,1746488025455000000,1746488037605000000,0
workspaceModel.updates.count,1746488025455000000,1746488037605000000,0
workspaceModel.updates.ms,1746488025455000000,1746488037605000000,0
workspaceModel.initializing.ms,1746488025455000000,1746488037605000000,0
workspaceModel.to.snapshot.ms,1746488025455000000,1746488037605000000,0
workspaceModel.loading.from.cache.ms,1746488025455000000,1746488037605000000,19
workspaceModel.init.bridges.ms,1746488025455000000,1746488037605000000,0
workspaceModel.pre.handlers.ms,1746488025455000000,1746488037605000000,0
workspaceModel.replace.project.model.ms,1746488025455000000,1746488037605000000,0
AWTEventQueue.dispatchTime90PNs,1746488025455000000,1746488037605000000,28159
AWTEventQueue.dispatchTimeAvgNs,1746488025455000000,1746488037605000000,369873.55549554416
NonBlockingReadAction.finalizedExecutionTimeUs,1746488025455000000,1746488037605000000,42395
WriteAction.executionsCount,1746488025455000000,1746488037605000000,4
AWTEventQueue.eventsDispatched,1746488025455000000,1746488037605000000,3703
FlushQueue.waitingTimeMaxNs,1746488025455000000,1746488037605000000,8321499135
FlushQueue.executionTimeMaxNs,1746488025455000000,1746488037605000000,27000831
FlushQueue.waitingTime90PNs,1746488025455000000,1746488037605000000,5771362303
FlushQueue.executionTimeAvgNs,1746488025455000000,1746488037605000000,187180.566439523
AWTEventQueue.dispatchTimeTotalNS,1746488025455000000,1746488037605000000,1368700411
AWTEventQueue.dispatchTimeMaxNs,1746488025455000000,1746488037605000000,884998143
FlushQueue.queueSize90P,1746488025455000000,1746488037605000000,1319
FlushQueue.executionTime90PNs,1746488025455000000,1746488037605000000,89599
NonBlockingReadAction.failedExecutionTimeUs,1746488025455000000,1746488037605000000,0
ReadAction.executionsCount,1746488025455000000,1746488037605000000,156995
NonBlockingReadAction.finalizedExecutionsCount,1746488025455000000,1746488037605000000,15
FlushQueue.queueSizeAvg,1746488025455000000,1746488037605000000,740.087734241908
NonBlockingReadAction.failedExecutionsCount,1746488025455000000,1746488037605000000,0
FlushQueue.queueSizeMax,1746488025455000000,1746488037605000000,1479
FlushQueue.waitingTimeAvgNs,1746488025455000000,1746488037605000000,2.0690473430187395E9
FlushQueue.tasksExecuted,1746488025455000000,1746488037605000000,2348
jps.global.get.libraries.ms,1746488025455000000,1746488037605000000,9
workspaceModel.mutableEntityStorage.replace.by.source.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.entities.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.resolve.ms,1746488025455000000,1746488037605000000,0
jps.library.entities.serializer.save.entities.ms,1746488025455000000,1746488037605000000,0
workspaceModel.moduleManagerBridge.build.module.graph.ms,1746488025455000000,1746488037605000000,0
jps.global.handle.before.change.events.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.entities.by.source.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.referrers.ms,1746488025455000000,1746488037605000000,0
jps.save.global.entities.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.to.snapshot.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.put.entity.ms,1746488025455000000,1746488037605000000,0
workspaceModel.global.updates.ms,1746488025455000000,1746488037605000000,0
jps.global.get.library.by.name.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.has.same.entities.ms,1746488025455000000,1746488037605000000,0
workspaceModel.moduleManagerBridge.newModule.ms,1746488025455000000,1746488037605000000,0
workspaceModel.load.cache.metadata.from.file.ms,1746488025455000000,1746488037605000000,0
workspaceModel.global.apply.state.to.project.builder.ms,1746488025455000000,1746488037605000000,0
workspaceModel.global.apply.state.to.project.ms,1746488025455000000,1746488037605000000,0
workspaceModel.save.cache.to.file.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.modify.entity.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.mutable.vfurl.index.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.add.entity.ms,1746488025455000000,1746488037605000000,0
workspaceModel.global.updates.count,1746488025455000000,1746488037605000000,0
jps.global.get.library.ms,1746488025455000000,1746488037605000000,0
workspaceModel.moduleManagerBridge.get.modules.ms,1746488025455000000,1746488037605000000,0
jps.library.entities.serializer.load.entities.ms,1746488025455000000,1746488037605000000,0
jps.global.initialize.library.bridges.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.add.diff.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.mutable.ext.mapping.ms,1746488025455000000,1746488037605000000,0
jps.app.storage.content.reader.load.component.ms,1746488025455000000,1746488037605000000,1
workspaceModel.moduleManagerBridge.new.nonPersistent.module.ms,1746488025455000000,1746488037605000000,0
jps.global.handle.changed.events.ms,1746488025455000000,1746488037605000000,0
workspaceModel.moduleManagerBridge.set.unloadedModules.ms,1746488025455000000,1746488037605000000,0
workspaceModel.moduleManagerBridge.load.module.ms,1746488025455000000,1746488037605000000,0
workspaceModel.sync.entities.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.collect.changes.ms,1746488025455000000,1746488037605000000,0
workspaceModel.mutableEntityStorage.remove.entity.ms,1746488025455000000,1746488037605000000,0
workspaceModel.moduleManagerBridge.create.module.instance.ms,1746488025455000000,1746488037605000000,0
workspaceModel.moduleManagerBridge.load.all.modules.ms,1746488025455000000,1746488037605000000,0
jps.load.initial.state.ms,1746488025455000000,1746488037605000000,5
jps.global.initialize.library.bridges.after.loading.ms,1746488025455000000,1746488037605000000,0
workspaceModel.load.cache.from.file.ms,1746488025455000000,1746488037605000000,0
FilePageCache.pageHits,1746488025455000000,1746488037605000000,0
FileChannelInterruptsRetryer.totalRetriedAttempts,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.housekeeperTimeSpentMs,1746488025455000000,1746488037605000000,38
FilePageCacheLockFree.totalBytesWritten,1746488025455000000,1746488037605000000,16673
FilePageCacheLockFree.totalPagesReclaimed,1746488025455000000,1746488037605000000,3
FilePageCache.capacityInBytes,1746488025455000000,1746488037605000000,503316478
FilePageCacheLockFree.housekeeperTurnsSkipped,1746488025455000000,1746488037605000000,9182
FilePageCacheLockFree.totalPagesReadMs,1746488025455000000,1746488037605000000,2
MappedFileStorage.totalPagesMapped,1746488025455000000,1746488037605000000,0
FilePageCache.totalPageLoadsUs,1746488025455000000,1746488037605000000,125518
FilePageCache.uncachedFileAccess,1746488025455000000,1746488037605000000,712
FilePageCache.pageLoadsAboveSizeThreshold,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.totalPagesWriteMs,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.housekeeperTurnsDone,1746488025455000000,1746488037605000000,26
FilePageCache.pageLoads,1746488025455000000,1746488037605000000,399
StorageLockContext.competingThreads.90P,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.totalNativeBytesAllocated,1746488025455000000,1746488037605000000,24576
FilePageCacheLockFree.totalPagesRequested,1746488025455000000,1746488037605000000,270
MappedFileStorage.totalBytesMapped,1746488025455000000,1746488037605000000,0
DirectByteBufferAllocator.totalSizeOfBuffersAllocatedInBytes,1746488025455000000,1746488037605000000,41967616
FilePageCache.maxRegisteredFiles,1746488025455000000,1746488037605000000,1097
FilePageCacheLockFree.totalPagesWritten,1746488025455000000,1746488037605000000,3
DirectByteBufferAllocator.disposed,1746488025455000000,1746488037605000000,356
FilePageCache.pageFastCacheHits,1746488025455000000,1746488037605000000,27800
DirectByteBufferAllocator.totalSizeOfBuffersCachedInBytes,1746488025455000000,1746488037605000000,41967616
FilePageCacheLockFree.totalBytesRead,1746488025455000000,1746488037605000000,0
StorageLockContext.competingThreads.avg,1746488025455000000,1746488037605000000,0.13043478260869565
DirectByteBufferAllocator.hits,1746488025455000000,1746488037605000000,3
DirectByteBufferAllocator.reclaimed,1746488025455000000,1746488037605000000,46
MappedFileStorage.totalTimeSpentOnMappingUs,1746488025455000000,1746488037605000000,674442
FilePageCacheLockFree.totalNativeBytesReclaimed,1746488025455000000,1746488037605000000,24576
FilePageCache.disposedBuffers,1746488025455000000,1746488037605000000,399
MappedFileStorage.storages,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.totalBytesRequested,1746488025455000000,1746488037605000000,2211840
DirectByteBufferAllocator.misses,1746488025455000000,1746488037605000000,399
FilePageCache.maxCacheSizeInBytes,1746488025455000000,1746488037605000000,415236096
FilePageCacheLockFree.totalClosedStoragesReclaimed,1746488025455000000,1746488037605000000,2
FilePageCacheLockFree.totalHeapBytesAllocated,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.totalPagesRequestsMs,1746488025455000000,1746488037605000000,7
StorageLockContext.competingThreads.max,1746488025455000000,1746488037605000000,3
FilePageCacheLockFree.nativeBytesInUse,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.totalPagesHandedOver,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.heapBytesInUse,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.totalPageAllocationsWaited,1746488025455000000,1746488037605000000,0
FilePageCacheLockFree.totalPagesAllocated,1746488025455000000,1746488037605000000,3
FilePageCacheLockFree.totalHeapBytesReclaimed,1746488025455000000,1746488037605000000,0
FilePageCache.totalPageDisposalsUs,1746488025455000000,1746488037605000000,3671
