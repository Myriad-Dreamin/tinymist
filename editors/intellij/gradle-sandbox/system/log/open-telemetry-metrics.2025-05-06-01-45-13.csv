# OpenTelemetry Metrics report: .csv, 4 fields: 
# <metric name>, <period start, nanoseconds>, <period end, nanoseconds>, <metric value>
# See CsvMetricsExporter for details.
# 
# NAME, PERIOD_START_NANOS, PERIOD_END_NANOS, VALUE
workspaceModel.sync.entities.ms,1746488713395000000,1746488759374000000,0
jps.global.get.libraries.ms,1746488713395000000,1746488759374000000,1
workspaceModel.mutableEntityStorage.modify.entity.ms,1746488713395000000,1746488759374000000,0
jps.global.get.library.ms,1746488713395000000,1746488759374000000,0
jps.global.get.library.by.name.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.has.same.entities.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.to.snapshot.ms,1746488713395000000,1746488759374000000,0
jps.save.global.entities.ms,1746488713395000000,1746488759374000000,38
workspaceModel.mutableEntityStorage.collect.changes.ms,1746488713395000000,1746488759374000000,0
jps.global.handle.before.change.events.ms,1746488713395000000,1746488759374000000,0
jps.global.initialize.library.bridges.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.add.entity.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.referrers.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.entities.by.source.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.mutable.ext.mapping.ms,1746488713395000000,1746488759374000000,0
workspaceModel.load.cache.metadata.from.file.ms,1746488713395000000,1746488759374000000,0
workspaceModel.global.apply.state.to.project.builder.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.add.diff.ms,1746488713395000000,1746488759374000000,0
workspaceModel.save.cache.to.file.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.remove.entity.ms,1746488713395000000,1746488759374000000,0
workspaceModel.global.apply.state.to.project.ms,1746488713395000000,1746488759374000000,0
jps.load.initial.state.ms,1746488713395000000,1746488759374000000,6
workspaceModel.mutableEntityStorage.entities.ms,1746488713395000000,1746488759374000000,1
jps.global.initialize.library.bridges.after.loading.ms,1746488713395000000,1746488759374000000,0
jps.app.storage.content.writer.save.component.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.mutable.vfurl.index.ms,1746488713395000000,1746488759374000000,0
workspaceModel.load.cache.from.file.ms,1746488713395000000,1746488759374000000,0
workspaceModel.global.updates.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.put.entity.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.resolve.ms,1746488713395000000,1746488759374000000,0
jps.library.entities.serializer.save.entities.ms,1746488713395000000,1746488759374000000,0
workspaceModel.global.updates.count,1746488713395000000,1746488759374000000,0
jps.global.handle.changed.events.ms,1746488713395000000,1746488759374000000,0
workspaceModel.mutableEntityStorage.replace.by.source.ms,1746488713395000000,1746488759374000000,0
jps.app.storage.content.reader.load.component.ms,1746488713395000000,1746488759374000000,1
jps.library.entities.serializer.load.entities.ms,1746488713395000000,1746488759374000000,0
NonBlockingReadAction.finalizedExecutionsCount,1746488713395000000,1746488759374000000,1
AWTEventQueue.dispatchTimeTotalNS,1746488713395000000,1746488759374000000,3712260030
NonBlockingReadAction.failedExecutionsCount,1746488713395000000,1746488759374000000,0
NonBlockingReadAction.finalizedExecutionTimeUs,1746488713395000000,1746488759374000000,127377
NonBlockingReadAction.failedExecutionTimeUs,1746488713395000000,1746488759374000000,0
DirectByteBufferAllocator.misses,1746488713395000000,1746488759374000000,403
FilePageCacheLockFree.housekeeperTurnsSkipped,1746488713395000000,1746488759374000000,34946
FilePageCacheLockFree.totalHeapBytesAllocated,1746488713395000000,1746488759374000000,0
StorageLockContext.competingThreads.max,1746488713395000000,1746488759374000000,0
FilePageCache.pageLoadsAboveSizeThreshold,1746488713395000000,1746488759374000000,0
FilePageCacheLockFree.housekeeperTurnsDone,1746488713395000000,1746488759374000000,94
FilePageCacheLockFree.totalPagesRequestsMs,1746488713395000000,1746488759374000000,4
FilePageCacheLockFree.totalPagesReadMs,1746488713395000000,1746488759374000000,1
FilePageCache.totalPageDisposalsUs,1746488713395000000,1746488759374000000,1744
FilePageCacheLockFree.totalBytesWritten,1746488713395000000,1746488759374000000,0
FilePageCacheLockFree.totalPageAllocationsWaited,1746488713395000000,1746488759374000000,0
FilePageCacheLockFree.totalPagesHandedOver,1746488713395000000,1746488759374000000,0
FilePageCacheLockFree.totalNativeBytesReclaimed,1746488713395000000,1746488759374000000,16384
FilePageCache.totalPageLoadsUs,1746488713395000000,1746488759374000000,170252
FilePageCacheLockFree.totalPagesRequested,1746488713395000000,1746488759374000000,4
FilePageCacheLockFree.totalHeapBytesReclaimed,1746488713395000000,1746488759374000000,0
FilePageCacheLockFree.totalBytesRead,1746488713395000000,1746488759374000000,8720
FilePageCache.pageHits,1746488713395000000,1746488759374000000,0
StorageLockContext.competingThreads.90P,1746488713395000000,1746488759374000000,0
FilePageCache.pageLoads,1746488713395000000,1746488759374000000,404
FilePageCache.maxRegisteredFiles,1746488713395000000,1746488759374000000,1102
FilePageCacheLockFree.heapBytesInUse,1746488713395000000,1746488759374000000,0
StorageLockContext.competingThreads.avg,1746488713395000000,1746488759374000000,0.0
FilePageCache.pageFastCacheHits,1746488713395000000,1746488759374000000,8800
FilePageCache.maxCacheSizeInBytes,1746488713395000000,1746488759374000000,418398208
FilePageCacheLockFree.totalPagesWriteMs,1746488713395000000,1746488759374000000,0
FilePageCache.disposedBuffers,1746488713395000000,1746488759374000000,404
FileChannelInterruptsRetryer.totalRetriedAttempts,1746488713395000000,1746488759374000000,0
FilePageCacheLockFree.totalNativeBytesAllocated,1746488713395000000,1746488759374000000,16384
FilePageCacheLockFree.totalBytesRequested,1746488713395000000,1746488759374000000,32768
DirectByteBufferAllocator.totalSizeOfBuffersCachedInBytes,1746488713395000000,1746488759374000000,41975808
FilePageCacheLockFree.totalPagesReclaimed,1746488713395000000,1746488759374000000,2
DirectByteBufferAllocator.reclaimed,1746488713395000000,1746488759374000000,47
FilePageCacheLockFree.totalPagesAllocated,1746488713395000000,1746488759374000000,2
FilePageCacheLockFree.nativeBytesInUse,1746488713395000000,1746488759374000000,0
FilePageCacheLockFree.totalClosedStoragesReclaimed,1746488713395000000,1746488759374000000,2
FilePageCacheLockFree.totalPagesWritten,1746488713395000000,1746488759374000000,0
FilePageCache.capacityInBytes,1746488713395000000,1746488759374000000,503316478
FilePageCacheLockFree.housekeeperTimeSpentMs,1746488713395000000,1746488759374000000,174
FilePageCache.uncachedFileAccess,1746488713395000000,1746488759374000000,239
DirectByteBufferAllocator.hits,1746488713395000000,1746488759374000000,3
DirectByteBufferAllocator.totalSizeOfBuffersAllocatedInBytes,1746488713395000000,1746488759374000000,41975808
DirectByteBufferAllocator.disposed,1746488713395000000,1746488759374000000,359
