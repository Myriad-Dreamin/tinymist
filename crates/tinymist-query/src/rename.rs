use log::debug;
use lsp_types::TextEdit;

use crate::{
    find_definition, find_references, prelude::*, syntax::get_deref_target,
    validate_renaming_definition, SemanticRequest,
};

/// The [`textDocument/rename`] request is sent from the client to the server to
/// ask the server to compute a workspace change so that the client can perform
/// a workspace-wide rename of a symbol.
///
/// [`textDocument/rename`]: https://microsoft.github.io/language-server-protocol/specification#textDocument_rename
#[derive(Debug, Clone)]
pub struct RenameRequest {
    /// The path of the document to request for.
    pub path: PathBuf,
    /// The source code position to request for.
    pub position: LspPosition,
    /// The new name to rename to.
    pub new_name: String,
}

impl SemanticRequest for RenameRequest {
    type Response = WorkspaceEdit;

    fn request(self, ctx: &mut AnalysisContext) -> Option<Self::Response> {
        let source = ctx.source_by_path(&self.path).ok()?;

        let offset = ctx.to_typst_pos(self.position, &source)?;
        let cursor = offset + 1;

        let ast_node = LinkedNode::new(source.root()).leaf_at(cursor)?;
        debug!("ast_node: {ast_node:?}", ast_node = ast_node);

        let deref_target = get_deref_target(ast_node, cursor)?;

        let lnk = find_definition(ctx, source.clone(), deref_target.clone())?;

        validate_renaming_definition(&lnk)?;

        let def_use = ctx.def_use(source.clone())?;
        let references = find_references(ctx, def_use, deref_target, ctx.position_encoding())?;

        let mut editions = HashMap::new();

        let def_loc = {
            let def_source = ctx.source_by_id(lnk.fid).ok()?;

            let span_path = ctx.path_for_id(lnk.fid).ok()?;
            let uri = path_to_url(&span_path).ok()?;

            let Some(range) = lnk.name_range else {
                log::warn!("rename: no name range");
                return None;
            };

            LspLocation {
                uri,
                range: ctx.to_lsp_range(range, &def_source),
            }
        };

        for i in (Some(def_loc).into_iter()).chain(references) {
            let uri = i.uri;
            let range = i.range;
            let edits = editions.entry(uri).or_insert_with(Vec::new);
            edits.push(TextEdit {
                range,
                new_text: self.new_name.clone(),
            });
        }

        // todo: conflict analysis
        Some(WorkspaceEdit {
            changes: Some(editions),
            ..Default::default()
        })
    }
}
