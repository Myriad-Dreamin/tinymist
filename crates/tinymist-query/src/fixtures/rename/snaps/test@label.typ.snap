---
source: crates/tinymist-query/src/rename.rs
expression: "JsonRepr::new_redacted(result, &REDACT_LOC)"
input_file: crates/tinymist-query/src/fixtures/rename/label.typ
---
{
 "changeAnnotations": {
  "Typst Rename Labels": {
   "description": "The language server searched the labels ambiguously",
   "label": "Typst Rename Labels",
   "needsConfirmation": true
  }
 },
 "documentChanges": [
  {
   "edits": [
    {
     "annotationId": "Typst Rename Labels",
     "newText": "new_name",
     "range": "6:21:6:33"
    }
   ],
   "textDocument": {
    "uri": "s0.typ",
    "version": null
   }
  }
 ]
}
